internal/graph/check.go

```diff
@@ -0,0 +1,485 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/openfga/openfga/internal/dispatch"
+	dispatchpb "github.com/openfga/openfga/pkg/proto/dispatch/v1"
+	"github.com/openfga/openfga/pkg/tuple"
+	serverErrors "github.com/openfga/openfga/server/errors"
+	"github.com/openfga/openfga/storage"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"golang.org/x/sync/errgroup"
+)
+
+var tracer = otel.Tracer("openfga/internal/graph/check")
+
+// ConcurrentChecker exposes a method to perform Check requests. A ConcurrentChecker
+// delegates subproblems to the provided dispatch.Check dispatcher.
+type ConcurrentChecker struct {
+	dispatcher dispatch.Check
+	datastore  storage.OpenFGADatastore
+}
+
+// NewConcurrentChecker creates an instance of a ConcurrentChecker.
+func NewConcurrentChecker(d dispatch.Dispatcher) *ConcurrentChecker {
+	return &ConcurrentChecker{dispatcher: d}
+}
+
+// Check performs a check of the provided request and context. It is assumed that the request
+// is validated beforehand and the provided rewrite is looked up beforehand by fetching the
+// type information from the authorization model.
+func (cc *ConcurrentChecker) Check(ctx context.Context, req *dispatchpb.DispatchCheckRequest, rewrite *openfgapb.Userset) (*dispatchpb.DispatchCheckResponse, error) {
+
+	reducerFunc := cc.reduce(ctx, req, rewrite)
+
+	resolved, err := any(ctx, []ReduceableCheckFunc{reducerFunc})
+	resolved.Metadata = addCallToResponseMetadata(resolved.Metadata)
+
+	return resolved, err
+}
+
+// dispatch dispatches the provided request to the internal dispatcher and yields a value on
+// the result channel if the subproblem is successfully resolved or an error otherwise.
+func (cc *ConcurrentChecker) dispatch(req *dispatchpb.DispatchCheckRequest) ReduceableCheckFunc {
+	return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+		result, err := cc.dispatcher.DispatchCheck(ctx, req)
+		resultChan <- result
+		return err
+	}
+}
+
+// reduce reduces the check request by breaking it into an appropriate subproblem that
+// is resolved by evaluating the underlying userset rewrite rule.
+func (cc *ConcurrentChecker) reduce(
+	ctx context.Context,
+	req *dispatchpb.DispatchCheckRequest,
+	rewrite *openfgapb.Userset,
```

Comment:
I don't love that we've named our userset rewrites (vernacular of the Zanzibar paper) as simply `Userset`. The term `Userset` has become massively overloaded depending on the context. And our definition of a `Userset` rewrite rule is not very recursive or uniform either. It'd be nice to move toward a definition more like the following:

```
message UsersetRewrite {
    oneof rewrite_operation {
        SetOperation union = 1;
        SetOperation intersection = 2;
        SetOperation exclusion = 3;
    }
}

message SetOperation {
    message Child {
        message This {}
        message Nil {}

        oneof child_type {
            This _this = 1;
            ComputedUserset computed_userset = 2;
            TupleToUserset tuple_to_userset = 3;
            UsersetRewrite userset_rewrite = 4;
            Nil _nil = 6;
        }
    }
}
```